File: java-workshop/part-2-light-mocks/src/main/java/service/RepositoryStockService.java
package service;

import domain.OrderItem;
import domain.Stock;
import domain.StockReservation;
import repository.StockRepository;

import java.util.ArrayList;
import java.util.List;

public class RepositoryStockService implements StockService {
    private final StockRepository stockRepository;

    public RepositoryStockService(StockRepository stockRepository) {
        this.stockRepository = stockRepository;
    }

    @Override
    public List<StockReservation> reserveStock(List<OrderItem> items) {
        List<StockReservation> reservations = new ArrayList<>();

        for (OrderItem item : items) {
            Stock stock = stockRepository.findByProductId(item.productId());
            if (stock != null && stock.availableQuantity() >= item.quantity()) {
                stockRepository.updateQuantity(item.productId(), stock.availableQuantity() - item.quantity());
                reservations.add(new StockReservation(item.productId(), item.quantity(), true));
            } else {
                reservations.add(new StockReservation(item.productId(), item.quantity(), false));
            }
        }

        return reservations;
    }

    @Override
    public void releaseStock(List<StockReservation> reservations) {
        for (StockReservation reservation : reservations) {
            if (!reservation.reserved()) {
                continue;
            }

            Stock stock = stockRepository.findByProductId(reservation.productId());
            if (stock != null) {
                stockRepository.updateQuantity(
                    reservation.productId(),
                    stock.availableQuantity() + reservation.quantity()
                );
            }
        }
    }
}

File: typescript-workshop/part-2-light-mocks/src/service/RepositoryStockService.ts
import { OrderItem } from '../domain/OrderItem';
import { StockReservation } from '../domain/StockReservation';
import { StockRepository } from '../repository/StockRepository';
import { StockService } from './StockService';

export class RepositoryStockService implements StockService {
  constructor(private readonly stockRepository: StockRepository) {}

  reserveStock(items: ReadonlyArray<OrderItem>): StockReservation[] {
    return items.map((item) => {
      const stock = this.stockRepository.findByProductId(item.productId);
      if (stock !== null && stock.availableQuantity >= item.quantity) {
        this.stockRepository.updateQuantity(item.productId, stock.availableQuantity - item.quantity);
        return new StockReservation(item.productId, item.quantity, true);
      }
      return new StockReservation(item.productId, item.quantity, false);
    });
  }

  releaseStock(reservations: StockReservation[]): void {
    for (const reservation of reservations) {
      if (reservation.reserved) {
        const stock = this.stockRepository.findByProductId(reservation.productId);
        if (stock !== null) {
          this.stockRepository.updateQuantity(reservation.productId, stock.availableQuantity + reservation.quantity);
        }
      }
    }
  }
}
